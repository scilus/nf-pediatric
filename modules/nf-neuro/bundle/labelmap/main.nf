process BUNDLE_LABELMAP {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(bundles), path(centroids)

    output:
    tuple val(meta), path("*_labels.nii.gz")    , emit: labels
    tuple val(meta), path("*_labels.trk")       , emit: labels_trk
    tuple val(meta), path("*_distances.nii.gz") , emit: distances
    tuple val(meta), path("*_distances.trk")    , emit: distances_trk
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def nb_points = task.ext.nb_points ? "--nb_pts ${task.ext.nb_points} ": ""
    def colormap = task.ext.colormap ? "--colormap ${task.ext.colormap} ": ""
    def new_labelling = task.ext.new_labelling ? "--new_labelling ": ""

    """
    bundles=(${bundles.join(" ")})
    centroids=(${centroids.join(" ")})

    for index in \${!bundles[@]};
        do
        ext=\${bundles[index]#*.}
        pos=\$((\$(echo \${bundles[index]} | grep -b -o __ | cut -d: -f1)+2))
        bname=\${bundles[index]:\$pos}
        bname=\$(basename \${bname} .\${ext})

        scil_bundle_label_map.py \${bundles[index]} \${centroids[index]} \
            tmp_out $nb_points $colormap $new_labelling -f

        mv tmp_out/labels_map.nii.gz ${prefix}__\${bname}_labels.nii.gz
        mv tmp_out/distance_map.nii.gz ${prefix}__\${bname}_distances.nii.gz
        mv tmp_out/labels.trk ${prefix}__\${bname}_labels.trk
        mv tmp_out/distance.trk ${prefix}__\${bname}_distances.trk
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_bundle_label_map.py -h

    bundles=(${bundles.join(" ")})
    centroids=(${centroids.join(" ")})

    for index in \${!bundles[@]};
        do
        ext=\${bundles[index]#*.}
        pos=\$((\$(echo \${bundles[index]} | grep -b -o __ | cut -d: -f1)+2))
        bname=\${bundles[index]:\$pos}
        bname=\$(basename \${bname} .\${ext})

        touch ${prefix}__\${bname}_labels.nii.gz
        touch ${prefix}__\${bname}_labels.trk
        touch ${prefix}__\${bname}_distances.nii.gz
        touch ${prefix}__\${bname}_distances.trk
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
