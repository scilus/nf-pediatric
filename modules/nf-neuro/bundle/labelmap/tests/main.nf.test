nextflow_process {

    name "Test Process BUNDLE_LABELMAP"
    script "../main.nf"
    process "BUNDLE_LABELMAP"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "bundle"
    tag "bundle/labelmap"

    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tractometry.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("bundle - labelmap") {
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/IFGWM.trk"),
                        file("\${test_data_directory}/IFGWM_uni_c_10.trk"),
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    file(process.out.labels.get(0).get(1)).name,
                    file(process.out.labels_trk.get(0).get(1)).name,
                    file(process.out.distances.get(0).get(1)).name,
                    file(process.out.distances_trk.get(0).get(1)).name
                ).match() }
            )
        }
    }
}
