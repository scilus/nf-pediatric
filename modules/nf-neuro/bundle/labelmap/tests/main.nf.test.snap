{
    "bundle - labelmap": {
        "content": [
            [
                "versions.yml:md5,c126a7422f280ed6bd664cc0f6636931"
            ],
            "test__IFGWM_labels.nii.gz",
            "test__IFGWM_labels.trk",
            "test__IFGWM_distances.nii.gz",
            "test__IFGWM_distances.trk"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-22T16:31:38.189887"
    }
}